### PR TITLE
Fix multiple_miners* tests

### PR DESCRIPTION
**Pull request by jferrant**
Fixes both multiple_miners and multiple_miners_with_nakamoto_blocks tests.

### CODE REVIEW

**Comment by hstove** [comment]
Unfortunately it looks like `multiple_miners` is failing on [c60e4a6](https://github.com/stacks-network/stacks-core/pull/5237/commits/c60e4a63c83c95646b1a7aa6a8d4191a62b8e821)

**Comment by jferrant** [comment]
> Unfortunately it looks like `multiple_miners` is failing on [c60e4a6](https://github.com/stacks-network/stacks-core/pull/5237/commits/c60e4a63c83c95646b1a7aa6a8d4191a62b8e821)

NEIN! haha taking another wack at it.

**Comment by hstove** [comment]
Looks like `multiple_miners` is passing now!

**Comment by jferrant** [comment]
> Looks like `multiple_miners` is passing now!

Yes i think it might have been failing before with that weird timeout with no logging thing that seems to randomly afflict tests. :/

File: stacks-signer/src/client/stacks_client.rs [hunk 1]
```diff
@@ -78,6 +78,7 @@ pub struct StacksClient {
 
 #[derive(Deserialize)]
 struct GetStackersErrorResp {
+    #[allow(dead_code)]
     err_type: String,
```
Comment (obycode) [inline stacks-signer/src/client/stacks_client.rs:82]: Seems like no one uses this now. Let's just delete it?

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs [hunk 2]
```diff
@@ -2133,6 +2133,7 @@ fn multiple_miners() {
     let node_2_p2p = 51025;
     let http_origin = format!("http://{}", &naka_conf.node.rpc_bind);
     naka_conf.miner.wait_on_interim_blocks = Duration::from_secs(1);
+    naka_conf.node.pox_sync_sample_secs = 1;
```
Comment (obycode) [inline testnet/stacks-node/src/tests/nakamoto_integrations.rs:2136]: What's the reason for this change (and other places where you set this to `1`? Looking at this config option, it seems strange that it's being used both as a time duration in seconds, and a number of samples... 🤔 (nothing you changed here). Should it default to 0 for these tests?

File: stacks-signer/src/client/stacks_client.rs (hunk 1 above)
Comment (jferrant) [reply stacks-signer/src/client/stacks_client.rs:82]: It will not deserialize correctly if i do that. Though I could deserialize the single field altrnatively.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs (hunk 2 above)
Comment (jferrant) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:2136]: any time there are multiple nodes, this needs to be a positive small number according to Jude (it may just be followers, but I was finding it an issue in here as well I think cauase of the bootstrapping). See https://nakamotoslack.slack.com/archives/C06N3DE9ARM/p1727029975879659?thread_ts=1726871759.956989&cid=C06N3DE9ARM

Comment (jferrant) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:2136]: I was seeing this test fail intermittently in CI prior to this change, so I assumed it fixed the issue, but I do not fully understand why.... :/ I know that is not a good answer though XD

File: testnet/stacks-node/src/tests/signer/mod.rs [hunk 3]
```diff
@@ -323,20 +323,16 @@ impl<S: Signer<T> + Send + 'static, T: SignerEventTrait + 'static> SignerTest<Sp
...
 
-        let t_start = Instant::now();
-        while test_observer::get_mined_nakamoto_blocks().is_empty() {
-            assert!(
-                t_start.elapsed() < timeout,
-                "Timed out while waiting for mined nakamoto block event"
-            );
-            thread::sleep(Duration::from_secs(1));
-        }
+        wait_for(timeout.as_secs(), || {
```
Comment (jferrant) [inline testnet/stacks-node/src/tests/signer/mod.rs:326]: This change is required because in multiple node set ups, there is only one test observer per node and you cannot guarantee which mines and therefore sees the block increment. This fixes a few tests.

File: testnet/stacks-node/src/tests/signer/mod.rs [hunk 4]
```diff
@@ -630,7 +626,7 @@ impl<S: Signer<T> + Send + 'static, T: SignerEventTrait + 'static> SignerTest<Sp
                     }
                 })
                 .collect::<Vec<_>>();
-            Ok(block_rejections.len() == expected_signers.len())
+            Ok(block_rejections.len() >= expected_signers.len())
```
Comment (jferrant) [inline testnet/stacks-node/src/tests/signer/mod.rs:629]: Unfortunately could not filter on a specific block hash so cannot properly filter the block responses hence the change to >= as we could have multiple rejections from earlier.

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 5]
```diff
@@ -1446,17 +1458,30 @@ fn multiple_miners() {
...
     } = run_loop_2.counters();
-    let _run_loop_2_thread = thread::Builder::new()
+    let run_loop_2_thread = thread::Builder::new()
         .name("run_loop_2".into())
         .spawn(move || run_loop_2.start(None, 0))
         .unwrap();
 
     signer_test.boot_to_epoch_3();
+
+    wait_for(120, || {
```
Comment (jferrant) [inline testnet/stacks-node/src/tests/signer/v0.rs:1474]: This was a precaution to ensure that both nodes have caught up to each other before proceeding. Not 100% sure it was necessary, but is safe to do and I think will prevent unintended race conditions.

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 6]
```diff
@@ -1553,6 +1578,12 @@ fn multiple_miners() {
         u64::try_from(miner_1_tenures + miner_2_tenures).unwrap()
     );
 
+    rl2_coord_channels
```
Comment (jferrant) [inline testnet/stacks-node/src/tests/signer/v0.rs:1581]: I was not 100% sure this was necessary either, but we were not properly terminating the runloop for the second node so I added it anyway.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs (hunk 2 above)
Comment (obycode) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:2136]: @jcnelson can you help clarify what number would be appropriate to use here. Also, while you're here, why is `pox_sync_sample_secs` used as both a number of seconds and a number of samples in the code? Is that intentional?

e.g., in the condition below, the first use of `self.max_samples` is as a number of samples, then the second use is as a number of seconds:
```rust
                    if self.watch_start_ts > 0
                        && (self.new_attachable_blocks.len() as u64) < self.max_samples
                        && self.watch_start_ts
                            + self.max_samples
                            + self.steady_state_burnchain_sync_interval
                                * (burnchain.stable_confirmations as u64)
                            < get_epoch_time_secs()
```

File: testnet/stacks-node/src/tests/signer/mod.rs (hunk 4 above)
Comment (kantai) [reply testnet/stacks-node/src/tests/signer/mod.rs:629]: I think something along the lines of what I do in this commit is better (https://github.com/stacks-network/stacks-core/pull/5238/commits/6a3746c9f6ca4edec8e9ce66efdbf074e87f5615) -- otherwise, you end up letting the same signer's rejections get counted multiple times because of the way the chunk events work.

**Review by hstove — APPROVED** [review]
LGTM!

**Review by kantai — APPROVED** [review]

**Comment by kantai** [comment]
This LGTM -- my only comment is that you should use the approach from https://github.com/stacks-network/stacks-core/commit/6a3746c9f6ca4edec8e9ce66efdbf074e87f5615 in the rejection tracking.

**Review by hstove — DISMISSED** [review]

**Comment by obycode** [comment]
This fixes `tests::signer::v0::partial_tenure_fork` too!

**Review by kantai — DISMISSED** [review]

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs (hunk 2 above)
Comment (jferrant) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs]: So to be clear, the pox_sync_sample_secs is used to handle an ambiguous state in epoch 2.x where the anchor block at the reward cycle boundary may not be known, i.e. it may have arrived or it may not have. It can use this config option as a precaution to enter a quiescent state which is a strong indicator that either the anchor block is known or it will never arrive (note it samples every one second hence why it is used both as a timer and also as a number of samples)

**Review by kantai — APPROVED** [review]

**Review by obycode — APPROVED** [review]
